src/v/bytes/iobuf.cc

```diff
@@ -222,3 +222,61 @@ bool iobuf::operator==(std::string_view o) const {
       });
     return are_equal;
 }
+
+/**
+ * For debugging, string-ize the iobuf in a format like "hexdump -C"
+ *
+ * This is useful if you are handling a parse error and would like to
+ * safely log the unparseable content.  Set an appropriate `limit` to avoid
+ * your log being too verbose.
+ *
+ * @param limit maximum number of bytes to read.
+ * @return a string populated with the following format:
+ *
+00000000 | 7b 22 76 65 72 73 69 6f 6e  22 3a 31 2c 22 6e 61  | {"version":1,"na
+00000010 | 6d 65 73 70 61 63 65 22 3a  22 74 65 73 74 2d 6e  | mespace":"test-n
+00000020 | 73 22 2c 22 74 6f 70 69 63  22 3a 22 74 65 73 74  | s","topic":"test
+00000030 | 2d 74 6f 70 69 63 22 2c 22  70 61 72 74 69 74 69  | -topic","partiti
+ */
+std::string iobuf::hexdump(size_t limit) const {
+    constexpr size_t line_length = 16;
+    auto result = std::ostringstream();
+    size_t total = 0;
+    std::string trail;
+    for (const auto& frag : *this) {
+        auto data = frag.get();
+        for (size_t i = 0; i < frag.size(); ++i) {
+            if (total % line_length == 0) {
+                if (trail.size()) {
+                    result << " | " << trail;
+                    trail.erase();
+                }
+                result << "\n  " << fmt::format("{:08x}", total) << " | ";
```

Comment:
nit: I feel like fmt::format_to could be slightly better, and it could be used directly with a std::string instead of a std::ostringstream